lib/AST/ASTPrinter.cpp

```diff
@@ -813,6 +814,18 @@ static StaticSpellingKind getCorrectStaticSpelling(const Decl *D) {
   }
 }
 
+static bool hasMutatingGetter(AbstractStorageDecl *ASD) {
+  return ASD->getGetter() && ASD->isGetterMutating();
+}
+
+static bool hasNonMutatingSetter(AbstractStorageDecl *ASD) {
+  if (!ASD->isSettable(nullptr)) return false;
+  auto ctxType = ASD->getDeclContext()->getDeclaredInterfaceType();
+  return !ASD->isSetterMutating() &&
+         ASD->isInstanceMember() &&
+         !ctxType->hasReferenceSemantics();
```

Comment:
@slavapestov, there's not really a better way to do this check that'll handle constrained protocol extensions, right?

(Also, the `ASD->isInstanceMember` check isn't quite right; if it's a non-instance member, it might still have a nonmutating setter. I think it's `!A && (!B || !C)`.)

Reply:
This is the same logic that was factored out of `PrintAST::printAccessors`, but I think you're right about the logic being wrong.

Reply:
Turns out several tests were relying on the existing behavior 😞